src/components/users/List.vue

```diff
@@ -0,0 +1,162 @@
+<script setup lang="ts">
+import { useAsyncState } from '@vueuse/core'
+import { ref, Ref } from 'vue'
+import { useAuth } from '../../composables/useAuth'
+import { useSessionToken } from '../../composables/useSessionToken'
+import { Users } from '../../lib/api/Users'
+import Paginator from '../leaderboards/Paginator.vue'
+import { UserRole } from '../../lib/api/data-contracts'
+
+const props = defineProps<{
+	limit: number | undefined
+	page: number
+	roles: string | string[] | undefined
+}>()
+
+const allRoles: UserRole[] = [
+	'Administrator',
+	'Banned',
+	'Confirmed',
+	'Registered'
+]
+
+function parseRolesFromQuery(roles: string | string[] | undefined): UserRole[] {
+	if (roles === undefined) {
+		return []
+	}
+
+	if (typeof roles === 'string') {
+		return allRoles.filter((role) => role === roles)
+	}
+
+	return allRoles.filter((r) => roles.includes(r))
+}
+
+const roles: Ref<UserRole[]> = ref(parseRolesFromQuery(props.roles))
+
+const userClient = new Users({
+	baseUrl: import.meta.env.VITE_BACKEND_URL
+})
+
+const token = useSessionToken()
+
+const {
+	state: users,
+	error,
+	isLoading,
+	execute
+} = useAsyncState(
+	async () => {
+		const resp = await userClient.listUsers(
+			{
+				// @ts-ignore The query param accepts a comma-separated list of roles,
+				// which is something the generated contract can't feasibly make types
+				// for - zysim
+				role: roles.value.length > 0 ? roles.value.join(',') : undefined,
+				limit: props.limit,
+				offset: (props.page - 1) * (props.limit ?? 0)
+			},
+			useAuth(token.value)
+		)
+
+		return resp.data
+	},
+	{
+		data: [],
+		limitDefault: 0,
+		limitMax: 0,
+		total: 0
+	}
+)
+</script>
+
+<template>
+	<div class="container">
+		<h1>Users</h1>
+		<form class="input-container">
+			<input hidden name="resultsPerPage" :value="props.limit" />
+			<input hidden name="page" :value="props.page" />
+			<label class="label" for="roles">Filter roles:</label>
+			<select id="roles" v-model="roles" name="role" class="input" multiple>
+				<option value="Administrator">Admin</option>
+				<option value="Registered">Registered</option>
+				<option value="Confirmed">Confirmed</option>
+				<option value="Banned">Banned</option>
+			</select>
+			<button class="button">Filter</button>
```

Comment:
Why not make this reactive, i.e. filter whenever the selection changes?

Reply:
You don't wanna do that. I tried; It's too "quick". You'd want to choose which options you really want first before committing.